src/main/java/org/prebid/server/bidder/adkerneladn/AdkernelAdnBidder.java

```diff
@@ -0,0 +1,256 @@
+
+        for (ExtImpAdkernelAdn impExt : pubToImps.keySet()) {
+            final BidRequest outgoingRequest = createBidRequest(prebidBidrequest, pubToImps.get(impExt));
+            final String body = Json.encode(outgoingRequest);
+            result.add(HttpRequest.of(HttpMethod.POST, buildEndpoint(impExt, endpointUrl), body, headers(),
+                    outgoingRequest));
+        }
+
+        return result;
+    }
+
+    private static BidRequest createBidRequest(BidRequest prebidBidRequest, List<Imp> imps) {
+        final BidRequest.BidRequestBuilder bidRequestBuilder = prebidBidRequest.toBuilder();
+
+        final List<Imp> modifiedImps = new ArrayList<>();
+        for (Imp imp : imps) {
+            final Imp modifiedImp = imp.toBuilder()
+                    .tagid(imp.getId())
+                    .build();
+
+            modifiedImps.add(modifiedImp);
+        }
+        bidRequestBuilder.imp(modifiedImps);
+
+        if (prebidBidRequest.getSite() != null) {
+            final Site modifySite = prebidBidRequest.getSite().toBuilder()
+                    .publisher(null).domain("").build();
+
+            bidRequestBuilder.site(modifySite);
+        }
+        if (prebidBidRequest.getApp() != null) {
+            final App modifyApp = prebidBidRequest.getApp().toBuilder()
+                    .publisher(null).build();
+
+            bidRequestBuilder.app(modifyApp);
+        }
+        return bidRequestBuilder.build();
+    }
+
+    // Builds enpoint url based on adapter-specific pub settings from imp.ext
+    private static String buildEndpoint(ExtImpAdkernelAdn impExt, String endpointUrl) {
+
+        if (impExt.getHost() != null) {
+            try {
+                final URL defaultUrl = new URL(endpointUrl);
+                final String defaultDomain = defaultUrl.getHost();
+                final String defaultPort = defaultUrl.getPort() == -1 ? "" : ":" + defaultUrl.getPort();
+
+                endpointUrl = endpointUrl.replace(defaultDomain + defaultPort, impExt.getHost());
+            } catch (MalformedURLException e) {
+                e.printStackTrace();
+            }
+        }
+
+        return String.format("%s%s", endpointUrl, impExt.getPubId());
+    }
+
+    @Override
+    public Result<List<BidderBid>> makeBids(HttpCall<BidRequest> httpCall, BidRequest bidRequest) {
+        try {
+            final BidResponse bidResponse = Json.decodeValue(httpCall.getResponse().getBody(), BidResponse.class);
+            return Result.of(extractBids(httpCall.getRequest().getPayload(), bidResponse), Collections.emptyList());
+        } catch (DecodeException | PreBidException e) {
+            return Result.emptyWithError(BidderError.badServerResponse(e.getMessage()));
+        }
+    }
+
+    private static List<BidderBid> extractBids(BidRequest bidRequest, BidResponse bidResponse) {
+        if (bidResponse == null || bidResponse.getSeatbid() == null) {
+            return Collections.emptyList();
+        } else {
```

Comment:
No need to use `else {...}` clause here, or extract return result as separate variable.